### PR TITLE
Add args to tree

### PR DESCRIPTION
**Pull request by rwwagner90**
Rebasing seemed hard, so opening a new PR

### CODE REVIEW

**Comment by rwwagner90** [comment]
@chancancode @nummi there were a lot of conflicts since the big PR was merged, so I cherry picked our commits into this new PR. Please take a look!

File: app/controllers/component-tree.js [hunk 1]
```diff
@@ -194,10 +194,18 @@ class RenderItem {
...
+  }
+
   get id() {
     return this.renderNode.id;
   }
 
+  get isCurlyInvocation() {
+    return this.renderNode && this.renderNode.args && this.renderNode.args.positional;
+  }
+
```
Comment (chancancode) [inline app/controllers/component-tree.js]: Minor but can you move both `get args` and `get isCurlyInvocation` to _after_ `get name`? (Trying to maintain roughly [this ordering](https://github.com/emberjs/ember-inspector/blob/master/ember_debug/libs/render-tree.js#L28-L38) in all places for consistency and make things slightly easier to find).

Also, don't need to check `this.renderNode` as it's guaranteed to be present.

Comment (chancancode) [inline app/controllers/component-tree.js]: Minor but can you move both `get args` and `get isCurlyInvocation` to _after_ `get name`? (Trying to maintain roughly [this ordering](https://github.com/emberjs/ember-inspector/blob/master/ember_debug/libs/render-tree.js#L28-L38) in all places for consistency and make things slightly easier to find).

Also, don't need to check `this.renderNode` as it's guaranteed to be present.

File: app/helpers/component-argument.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * Determine the type of the component argument for display
+ *
+ * @method componentArgumentDisplay
+ * @param {*} argument
+ * @return {*} The argument with the correct type for display
+ */
+export function componentArgumentDisplay([argument]) {
+  if (typeof argument === 'string') {
+    return JSON.stringify(argument).replace(/^\"|\"$/g, '');
```
Comment (chancancode) [inline app/helpers/component-argument.js]:
```suggestion
    // Escape any interior quotes – we will add the surrounding quotes in the template
    return argument.replace(/\"/g, '\\"');
```

File: app/styles/component_tree.scss [hunk 3]
```diff
@@ -7,6 +7,18 @@
...
+    background-color: var(--base05);
+
+    .component-tree-item__action {
+      opacity: 1;
+    }
+
+    .component-tree-item__action.disabled {
+      opacity: 0.65;
+    }
+  }
```
Comment (chancancode) [inline app/styles/component_tree.scss]: Remove this hunk (it was moved down due to linter complaints)

File: app/styles/component_tree.scss [hunk 4]
```diff
@@ -45,36 +57,70 @@
...
-  .component-tree-item__action {
-    opacity: 1;
-  }
+.component-tree-item-classic__bracket:before {
+  content: '{{';
+}
 
-  .component-tree-item__action.disabled {
-    opacity: 0.65;
-  }
```
Comment (chancancode) [inline app/styles/component_tree.scss]: Add this hunk back (it was moved down from above due to linter complaints)

File: app/templates/components/component-tree-item.hbs [hunk 5]
```diff
@@ -21,10 +20,81 @@
...
+        "component-tree-item__bracket"
+      }}
+      {{if
+        (and (not @item.isCurlyInvocation) (not @item.hasChildren) @item.isComponent)
+        "component-tree-item__bracket--self-closing"
+      }}
+      {{if
+        (and @item.isCurlyInvocation @item.isComponent)
+        "component-tree-item-classic__bracket"
+      }}"
```
Comment (chancancode) [inline app/templates/components/component-tree-item.hbs]:
```hbs
{{if @item.isComponent
  (if @item.isCurlyInvocation
    "component-tree-item-classic__bracket"
    (if @item.hasChildren
      "component-tree-item__bracket--self-closing"
      "component-tree-item__bracket"
    )
  )
}}
```

(you can indent it however the linter wants it, but we should group the conditions)

File: app/templates/components/component-tree-item.hbs [hunk 6]
```diff
@@ -21,10 +20,81 @@
...
+
+        {{#each-in @item.args.named as |name value|}}
+          <div class="arg-token">
+            {{#if (is-string value)}}
+              <span class="at-token">
+                @
+              </span>
+              <span class="key-token">
+                {{name}}
+              </span>
```
Comment (chancancode) [inline app/templates/components/component-tree-item.hbs]:
```
              <span class="key-token">
                 @{{name}}
               </span>
```

IMO we should just treat `@` as part of the key here.

File: app/templates/components/component-tree-item.hbs [hunk 7]
```diff
@@ -21,10 +20,81 @@
...
+              ="
+              <span class="value-token">
+                {{component-argument value}}
+              </span>
+              "
+            {{else}}
+              <span class="at-token">
+                @
+              </span>
```
Comment (chancancode) [inline app/templates/components/component-tree-item.hbs]: Same re: treating `@` as part of the key.

File: app/templates/components/component-tree-item.hbs [hunk 8]
```diff
@@ -21,10 +20,81 @@
...
+      {{#if @item.args.positional}}
+        <span class="component-name">
+          {{@item.name}}
+        </span>
+
+        {{#each @item.args.positional as |value|}}
+          <div class="arg-token">
+            <span class="value-token">
+              {{component-argument value}}
+            </span>
```
Comment (chancancode) [inline app/templates/components/component-tree-item.hbs]: This needs to be quoted now

```hbs
{{#if (is-string value)}}"{{/if}}
<span class="value-token">
  {{component-argument value}}
</span>
{{#if (is-string value)}}"{{/if}}
```

File: app/templates/components/component-tree-item.hbs [hunk 9]
```diff
@@ -21,10 +20,81 @@
...
+
+        {{#each-in @item.args.named as |name value|}}
+          <div class="arg-token">
+            <span class="key-token">
+              {{name}}
+            </span>
+            =
+            <span class="value-token">
+              {{component-argument value}}
+            </span>
```
Comment (chancancode) [inline app/templates/components/component-tree-item.hbs]:
```hbs
{{#if (is-string value)}}"{{/if}}
<span class="value-token">
  {{component-argument value}}
</span>
{{#if (is-string value)}}"{{/if}}
```

File: app/templates/components/component-tree-item.hbs [hunk 10]
```diff
@@ -21,10 +20,81 @@
...
+          </div>
+        {{/each-in}}
+      {{else}}
+        <span class="component-name">
+          {{classify @item.name}}
+        </span>
+
+        {{#each-in @item.args.named as |name value|}}
+          <div class="arg-token">
+            {{#if (is-string value)}}
```
Comment (chancancode) [inline app/templates/components/component-tree-item.hbs]: Instead of breaking up and duplicating the whole block, I think you can just add something like...

```hbs
{{#if (is-string value)}}"{{else}}<span class="bracket-token">\{{</span>{{/if}}
<span class="value-token">
  {{component-argument value}}
</span>
{{#if (is-string value)}}"{{else}}<span class="bracket-token">}}</span>{{/if}}
```

...around the values, similar to what I showed above.

**Comment by chancancode** [comment]
I left some review comments, but I think the main issue is still the horizontal scrolling CSS issues? (scrolling breaks the row highlight/closing tag/buttons)

File: app/helpers/component-argument.js (hunk 2 above)
Comment (rwwagner90) [reply app/helpers/component-argument.js]: @chancancode doesn't replacing all interior quotes negate what we were trying to do with `JSON.stringify` in the first place?

File: app/styles/component_tree.scss (hunk 3 above)
Comment (rwwagner90) [reply app/styles/component_tree.scss]: I removed the one that was moved down. This is now the only copy of this.

File: app/helpers/component-argument.js (hunk 2 above)
Comment (chancancode) [reply app/helpers/component-argument.js]: It is replacing the interior quotes with `\"` so it's escaping them. `JSON.stringify` does (basically) two things for us here: 1) escape the interior quotes, 2) add the surrounding quotes. We wanted both before. Now we only want (1) because we are just going to add the surrounding quotes in the template so that we can highlight them differently. So, now `JSON.stringify` is essentially doing too much for us and instead of running (1) + (2) and then reverting the effect of (2), we should just write the code to do (1) at this point, which is what my suggestion does.

File: app/styles/component_tree.scss (hunk 3 above)
Comment (chancancode) [reply app/styles/component_tree.scss]: Weren't the linter unhappy with it? (Something something descending specificity, seen in the `ember build` output)

File: app/controllers/component-tree.js (hunk 1 above)
Comment (rwwagner90) [reply app/controllers/component-tree.js]: @chancancode I would disagree with that ordering, in favor of alphabetizing I think. People are not going to be familiar with what the API returns, that I know of.

Comment (chancancode) [reply app/controllers/component-tree.js]: It's a logical grouping... `isRoot` and `isRouteTemplate` may sound similar but they do pretty different things and doesn't make a lot of sense to sort them next to each other just because of the name happens to match (and then you would have to resort when you refactor and change the name?). A lot of the other files (here, capture-render-tree, render-tree, tests, etc) have this same logical structure/ordering now, roughly from "most important" to "less important" things (subjective, but the ordering is at least consistent everywhere) and I would like to keep it that way. I just liked to that documentation comment because it happens to have enumerated them, but the code structure in other places follows that order (I try to at least).

File: app/templates/components/component-tree-item.hbs [hunk 11]
```diff
@@ -21,10 +20,75 @@
...
+    class="component-tree-item__tag
+      {{if
+        @item.isComponent
+        (if
+          @item.isCurlyInvocation
+          "component-tree-item-classic__bracket"
+          (if
+            @item.hasChildren
+            "component-tree-item__bracket--self-closing"
+            "component-tree-item__bracket"
```
Comment (chancancode) [inline app/templates/components/component-tree-item.hbs]: Oops. My bad, these last two are in the reverse order I think (if hasChildren -> *not* self-closing)

**Comment by nummi** [comment]
> I left some review comments, but I think the main issue is still the horizontal scrolling CSS issues? 

I don't know if we'll fix this before the next release. It has been an issue for a while and I'm afraid the fix will require some big changes to the markup and/or the CSS.

**Comment by chancancode** [comment]
@nummi I don't think the problem is specific to our setup here, you can reproduce it with a pretty minimal html/css file:

```html
<style>
.scrollable {
  overflow: scroll;
  width: 100%;
  height: 100%;
}

.row {
  background-color: lightgrey;
  height: 50px;
}

.row:nth-child(2n) {
  background-color: lightslategrey;
}
</style>

<div class="scrollable">
  <div class="row"></div><div class="row"></div><div class="row"></div><div class="row"></div><div class="row"></div><div class="row"></div><div class="row"></div><div class="row"></div><div class="row"></div><div class="row"></div>

  <div class="row" style="width: 5000px"></div>

  <div class="row"></div><div class="row"></div><div class="row"></div><div class="row"></div><div class="row"></div><div class="row"></div><div class="row"></div><div class="row"></div><div class="row"></div><div class="row"></div>
</div>
```

I think we may have to switch to flexbox, or perhaps there is some other CSS Trick™ that would make it work here, but I'm pretty sure we can make _something_ work. Since it's a pretty major new feature, it's okay for it to take a bit longer (the 3.13.0 release already went out yesterday, so it's not a blocker), but we should find a way to introduce it without introducing a regression.

**Comment by rwwagner90** [comment]
@chancancode I was talking to @nummi earlier and I think we should move the action icons into the object inspector. That would solve the issue of them being off the screen.

**Comment by rwwagner90** [comment]
@chancancode I believe I resolved all the things you requested. Mind taking another look? I think we should get this in, and then have @nummi help us move the icons to the object inspector, so people do not have to scroll to see them.

**Comment by nummi** [comment]
image

**Comment by rwwagner90** [comment]
Nice work @nummi!

**Comment by nummi** [comment]
https://github.com/emberjs/ember-inspector/pull/1129

File: app/controllers/component-tree.js [hunk 12]
```diff
@@ -218,6 +218,14 @@ class RenderItem {
     return this.renderNode.name;
   }
 
+  get args() {
+    return this.renderNode.args;
+  }
+
+  get isCurlyInvocation() {
+    return this.renderNode && this.renderNode.args && this.renderNode.args.positional;
```
Comment (chancancode) [inline app/controllers/component-tree.js]:
```suggestion
    return this.renderNode.args && this.renderNode.args.positional;
```

`this.renderNode` is guaranteed

File: app/templates/components/component-tree-item.hbs [hunk 13]
```diff
@@ -1,54 +1,111 @@
 <div
   style={{@item.style}}
   class="
-    component-tree-item
+    component-tree-item relative flex items-center mx-1 rounded
+    {{if @item.hasInstance "cursor-pointer" "cursor-default"}}
     {{if @item.isComponent "component-tree-item--component"}}
-    {{if @item.hasInstance "component-tree-item--has-instance"}}
-    {{if @item.isSelected "component-tree-item--selected"}}
-    {{if @item.isHighlighted "component-tree-item--highlighted"}}
```
Comment (chancancode) [inline app/templates/components/component-tree-item.hbs]: why is this removed?

File: app/templates/components/component-tree-item.hbs [hunk 14]
```diff
@@ -1,54 +1,111 @@
...
+              "component-tree-item__bracket"
+              "component-tree-item__bracket component-tree-item__bracket--self-closing"
+            )
+          )
+        }}"
     >
-      {{svg-jar "eye" width="20px" height="20px"}}
-    </button>
+      {{#if @item.isComponent}}
+        {{#if @item.args.positional}}
```
Comment (chancancode) [inline app/templates/components/component-tree-item.hbs]:
```suggestion
        {{#if @item.isCurlyInvocation}}
```

File: app/templates/components/component-tree-item.hbs [hunk 15]
```diff
@@ -1,54 +1,111 @@
...
-      {{on "click" @item.inspectElement}}
-    >
-      {{svg-jar "code-line" width="20px" height="20px"}}
-    </button>
+          {{#each @item.args.positional as |value|}}
+            <div class="arg-token flex ml-2">
+              <span class="value-token">
+                {{component-argument value}}
+              </span>
+            </div>
```
Comment (chancancode) [inline app/templates/components/component-tree-item.hbs]: This need to check for strings and add quotes.

File: app/templates/components/component-tree-item.hbs [hunk 16]
```diff
@@ -1,54 +1,111 @@
...
+
+          {{#each-in @item.args.named as |name value|}}
+            <div class="arg-token flex ml-2">
+              <span class="key-token">
+                {{name}}
+              </span>
+              =
+              <span class="value-token">
+                {{component-argument value}}
+              </span>
```
Comment (chancancode) [inline app/templates/components/component-tree-item.hbs]: This need to check for strings and add quotes.

**Comment by chancancode** [comment]
> args in Glimmer components and no args on classic

@nummi are you testing both on the same app? I think we expect args to show up for both, assuming you have a recent enough version of `ember-source`. So if it's not showing up for classic components, there may be a bug.

File: app/templates/components/component-tree-item.hbs (hunk 13 above)
Comment (nummi) [reply app/templates/components/component-tree-item.hbs]: As far as I could tell that was being used for the hover state. See: https://github.com/emberjs/ember-inspector/pull/1129/files#diff-cbd2aa0d940eef1f44430e6ca030fbd9L8

**Comment by nummi** [comment]
> @nummi are you testing both on the same app? I think we expect args to show up for both, assuming you have a recent enough version of ember-source. So if it's not showing up for classic components, there may be a bug.

I am not. The classic app I tested in is deprecations.emberjs.com and the Glimmer app is ember-source: 3.15.0

File: app/templates/components/component-tree-item.hbs (hunk 13 above)
Comment (nummi) [reply app/templates/components/component-tree-item.hbs]: Here is the original PR: https://github.com/emberjs/ember-inspector/pull/984

**Comment by chancancode** [comment]
@nummi 👌 that's fine then. if you have any classic components in the 3.15 app, they would/should also show up with args.

Another thing I noticed is sometimes the strings passed to the components can be *very* long. Do we want to come up with a reasonable limit and truncate the rest with `...`? (The `component-argument` can handle this pretty naturally.)

File: app/templates/components/component-tree-item.hbs (hunk 13 above)
Comment (chancancode) [reply app/templates/components/component-tree-item.hbs]: I see. Yeah that changed in Octane. It is now used to highlight the "currently previewing component". IMO the previous behavior is not that useful, since you can already see the parent/child relationship from the nesting.

There are two ways to "preview" a component, by hovering the rows in the component tree or by using the "live inspection" feature and hover over the DOM elements on the page:

highlight

Using CSS hover state only solves the first part and I think we still need this for the latter.

Comment (nummi) [reply app/templates/components/component-tree-item.hbs]: Oooooh. I see. I'll have to revert that.

**Comment by rwwagner90** [comment]
@chancancode coming up with a limit sounds good.

**Comment by nummi** [comment]
https://github.com/emberjs/ember-inspector/pull/1135

**Comment by chancancode** [comment]
> @chancancode coming up with a limit sounds good.

Feel free to steal this:

```js
function truncate(str, limit = 20, rest = "...") {
  if (str.length <= limit) {
    return str;
  }

  if (rest.length >= limit) {
    throw new Error("rest cannot be longer than limit");
  }

  let parts = str.split(/(\s+|\b)/); // split by whitespace or word boundaries
  let selected = []; // the parts to keep
  let targetLength = limit - rest.length; // leave room for the "..."
  let currentLength = 0;

  while (true) {
    let candidate = parts.shift();

    if ((currentLength + candidate.length) <= targetLength) {
      selected.push(candidate);
      currentLength += candidate.length;
    } else {
      if (currentLength === 0) {
        // We have no choice but to break in the middle of a long word
        selected.push(candidate.slice(0, targetLength));
      }

      break;
    }
  }

  selected.push(rest);

  return selected.join("")
}
```

It ensures the resulting string (including the "..." if added) is shorter than the given limit, but try to break at word boundaries or whitespaces when possible:

```
> truncate("It ensures the resulting string (including the \"...\" if added) is shorter than the given limit, but try to break at word boundaries or whitespaces when possible:", 999)
"It ensures the resulting string (including the "..." if added) is shorter than the given limit, but try to break at word boundaries or whitespaces when possible:"

> truncate("It ensures the resulting string (including the \"...\" if added) is shorter than the given limit, but try to break at word boundaries or whitespaces when possible:", 100)
"It ensures the resulting string (including the "..." if added) is shorter than the given limit, ..."

> truncate("It ensures the resulting string (including the \"...\" if added) is shorter than the given limit, but try to break at word boundaries or whitespaces when possible:", 50)
"It ensures the resulting string (including the ..."

> truncate("It ensures the resulting string (including the \"...\" if added) is shorter than the given limit, but try to break at word boundaries or whitespaces when possible:", 20)
"It ensures the ..."

> truncate("It ensures the resulting string (including the \"...\" if added) is shorter than the given limit, but try to break at word boundaries or whitespaces when possible:", 10)
"It ..."
```

With this style the truncation, I find that 20 seems to be a pretty reasonable default (it will often be much shorter than that depending the length of the word that happen to sit at the boundary).

**Comment by nummi** [comment]
Any chance we could use an ellipsis in place of three periods? "…"

**Comment by chancancode** [comment]
Sounds good to me. The algorithm shot is work just as well

**Comment by rwwagner90** [comment]
@chancancode I just pushed up the truncate and the adding quotes you requested. I was unable to test out the positional params because it seems like the renderNode is coming back with named args only, not positional ones, even when I use `{{set-body-class this.theme}}`
